ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/PreconditionChecker.java

```diff
@@ -0,0 +1,220 @@
+    this.serverState = Objects.requireNonNull(serverState, "serverState");
+    this.topicClient = new KafkaTopicClientImpl(
+        () -> createCommandTopicAdminClient(
+            new KsqlRestConfig(propertiesLoader.get()), new KsqlConfig(propertiesLoader.get())));
+    this.preconditions = restConfig.getConfiguredInstances(
+        KsqlRestConfig.KSQL_SERVER_PRECONDITIONS,
+        KsqlServerPrecondition.class
+    );
+    this.vertx = Vertx.vertx(
+        new VertxOptions()
+            .setMaxWorkerExecuteTimeUnit(TimeUnit.MILLISECONDS)
+            .setMaxWorkerExecuteTime(Long.MAX_VALUE));
+    this.server = new PreconditionServer(
+        vertx,
+        restConfig,
+        serverState
+    );
+  }
+
+  @VisibleForTesting
+  PreconditionChecker(
+      final Supplier<Map<String, String>> propertiesLoader,
+      final ServiceContext serviceContext,
+      final KsqlRestConfig restConfig,
+      final KafkaTopicClient topicClient,
+      final Vertx vertx,
+      final List<KsqlServerPrecondition> preconditions,
+      final PreconditionServer server,
+      final ServerState state
+  ) {
+    this.propertiesLoader = Objects.requireNonNull(propertiesLoader, "propertiesLoader");
+    this.serviceContext = Objects.requireNonNull(serviceContext, "serviceContext");
+    this.restConfig = Objects.requireNonNull(restConfig, "restConfig");
+    this.topicClient = Objects.requireNonNull(topicClient, "topicClient");
+    this.vertx = Objects.requireNonNull(vertx, "vertx");
+    this.preconditions = Objects.requireNonNull(preconditions, "preconditions");
+    this.server = Objects.requireNonNull(server, "server");
+    this.serverState = Objects.requireNonNull(state, "state");
+  }
+
+  private boolean shouldCheckPreconditions() {
+    return preconditions.stream()
+        .map(p -> p.checkPrecondition(propertiesLoader.get(), serviceContext, topicClient))
+        .peek(
+            r -> r.ifPresent(rr -> LOG.info("Precondition failed: {}", rr))
+        )
+        .anyMatch(Optional::isPresent);
+  }
+
+  @Override
+  public void startAsync() {
+    if (!shouldCheckPreconditions()) {
+      LOG.info("All preconditions passed, skipping precondition server start");
+      return;
+    }
+    LOG.info("Some preconditions not passed, starting precondition server");
+    server.start();
+  }
+
+  @Override
+  public void notifyTerminated() {
+    terminatedFuture.complete(null);
+  }
+
+  @Override
+  public void shutdown() {
+    if (server.started())
+      server.stop();
+    vertx.close();
+  }
+
+  @Override
+  public void awaitTerminated() {
+    final List<Predicate<Exception>> predicates = ImmutableList.of(
+        e -> !(e instanceof KsqlFailedPrecondition)
+    );
+    RetryUtil.retryWithBackoff(
+        Integer.MAX_VALUE,
+        1000,
+        30000,
+        this::checkPreconditions,
+        terminatedFuture::isDone,
+        predicates
+    );
+  }
+
+  private void checkPreconditions() {
```

Comment:
Could you add a comment on how this function continuously runs and determines if PreconditionServer will continue running or not? And if it passes, we'll terminate the pre condition server and move onto the main KSQL server.